File: Structure_demo/server/src/services/nodeManager.ts
/**
 * Node Manager Service
 * 
 * Manages the cluster of storage nodes:
 * - Tracks node health via heartbeats
 * - Distributes chunks across nodes
 * - Handles node failures and re-replication
 * - Provides node selection for uploads
 */

// Use native fetch or fallback
const fetchFn = globalThis.fetch || require('node-fetch');
import { Shard } from './erasureService';

// Configuration
const HEARTBEAT_INTERVAL = 10000;  // 10 seconds
const NODE_TIMEOUT = 30000;        // 30 seconds before marking offline
const REPLICATION_DELAY = 5000;    // Wait before starting re-replication

export interface StorageNode {
    id: string;
    url: string;
    port: number;
    status: 'online' | 'offline' | 'degraded';
    lastHeartbeat: Date;
    chunksStored: number;
    storageUsed: number;
    latency: number;        // ms
}

export interface ChunkPlacement {
    chunkId: string;
    shardIndex: number;
    nodeId: string;
    nodeUrl: string;
    stored: boolean;
    verified: boolean;
}

/**
 * Node Manager - Orchestrates storage cluster
 */
class NodeManager {
    private nodes: Map<string, StorageNode> = new Map();
    private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
    private listeners: ((event: string, data: any) => void)[] = [];

    constructor() {
        console.log('[NodeManager] Initialized');
    }

    /**
     * Register a storage node
     */
    registerNode(id: string, url: string, port: number): void {
        const node: StorageNode = {
            id,
            url,
            port,
            status: 'offline',
            lastHeartbeat: new Date(0),
            chunksStored: 0,
            storageUsed: 0,
            latency: 0
        };
        this.nodes.set(id, node);
        console.log(`[NodeManager] Registered node: ${id} at ${url}:${port}`);
    }

    /**
     * Register default nodes for development
     */
    registerDefaultNodes(): void {
        // Register 6 storage nodes (for RS(4,2) we need at least 6 nodes)
        const basePort = 4001;
        for (let i = 0; i < 6; i++) {
            const port = basePort + i;
            const id = `node-${port}`;
            this.registerNode(id, 'http://localhost', port);
        }
    }

    /**
     * Start monitoring node health
     */
    startMonitoring(): void {
        if (this.heartbeatTimer) {
            clearInterval(this.heartbeatTimer);
        }

        console.log('[NodeManager] Starting health monitoring');
        this.checkAllNodes();

        this.heartbeatTimer = setInterval(() => {
            this.checkAllNodes();
        }, HEARTBEAT_INTERVAL);
    }

    /**
     * Stop monitoring
     */
    stopMonitoring(): void {
        if (this.heartbeatTimer) {
            clearInterval(this.heartbeatTimer);
            this.heartbeatTimer = null;
        }
    }

    /**
     * Check health of all nodes
     */
    private async checkAllNodes(): Promise<void> {
        const checks = Array.from(this.nodes.values()).map(node =>
            this.checkNode(node)
        );
        await Promise.allSettled(checks);
    }

    /**
     * Check single node health
     */
    private async checkNode(node: StorageNode): Promise<void> {
        const startTime = Date.now();

        try {
            const response = await fetch(`${node.url}:${node.port}/health`, {
                method: 'GET',
                timeout: 5000
            } as any);

            if (response.ok) {
                const data = await response.json() as any;
                const latency = Date.now() - startTime;

                const wasOffline = node.status === 'offline';

                node.status = 'online';
                node.lastHeartbeat = new Date();
                node.chunksStored = data.chunksStored || 0;
                node.storageUsed = data.storageUsed || 0;
                node.latency = latency;

                if (wasOffline) {
                    console.log(`[NodeManager] Node ${node.id} came online (${latency}ms)`);
                    this.emit('node:online', { nodeId: node.id });
                }
            } else {
                this.markNodeOffline(node);
            }
        } catch (error) {
            this.markNodeOffline(node);
        }
    }

    /**
     * Mark a node as offline
     */
    private markNodeOffline(node: StorageNode): void {
        if (node.status !== 'offline') {
            console.warn(`[NodeManager] Node ${node.id} is OFFLINE`);
            node.status = 'offline';
            this.emit('node:offline', { nodeId: node.id });
        }
    }

    /**
     * Get list of online nodes
     */
    getOnlineNodes(): StorageNode[] {
        return Array.from(this.nodes.values()).filter(n => n.status === 'online');
    }

    /**
     * Get all nodes
     */
    getAllNodes(): StorageNode[] {
        return Array.from(this.nodes.values());
    }

    /**
     * Select nodes for storing shards
     * Uses round-robin with least-loaded preference
     */
    selectNodesForShards(shardCount: number): StorageNode[] {
        const online = this.getOnlineNodes();

        if (online.length < shardCount) {
            throw new Error(`Not enough online nodes: need ${shardCount}, have ${online.length}`);
        }

        // Sort by storage used (least loaded first) then by latency
        const sorted = [...online].sort((a, b) => {
            if (a.storageUsed !== b.storageUsed) {
                return a.storageUsed - b.storageUsed;
            }
            return a.latency - b.latency;
        });

        return sorted.slice(0, shardCount);
    }

    /**
     * Store a shard on a specific node
     */
    async storeShard(node: StorageNode, shard: Shard): Promise<ChunkPlacement> {
        const url = `${node.url}:${node.port}/chunks/${shard.id}`;

        try {
            const response = await fetch(url, {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/octet-stream'
                },
                body: new Uint8Array(shard.data)
            });

            if (!response.ok) {
                throw new Error(`Failed to store shard: ${response.statusText}`);
            }

            console.log(`[NodeManager] Stored shard ${shard.id.substring(0, 8)}... on ${node.id}`);

            return {
                chunkId: shard.id,
                shardIndex: shard.index,
                nodeId: node.id,
                nodeUrl: `${node.url}:${node.port}`,
                stored: true,
                verified: true
            };
        } catch (error) {
            console.error(`[NodeManager] Failed to store shard on ${node.id}:`, error);
            throw error;
        }
    }

    /**
     * Retrieve a shard from a node
     */
    async retrieveShard(nodeUrl: string, shardId: string): Promise<Buffer> {
        const url = `${nodeUrl}/chunks/${shardId}`;

        try {
            const response = await fetch(url);

            if (!response.ok) {
                throw new Error(`Failed to retrieve shard: ${response.statusText}`);
            }

            const arrayBuffer = await response.arrayBuffer();
            return Buffer.from(arrayBuffer);
        } catch (error) {
            console.error(`[NodeManager] Failed to retrieve shard ${shardId}:`, error);
            throw error;
        }
    }

    /**
     * Delete a shard from a node
     */
    async deleteShard(nodeUrl: string, shardId: string): Promise<void> {
        const url = `${nodeUrl}/chunks/${shardId}`;

        try {
            const response = await fetch(url, { method: 'DELETE' });

            if (!response.ok && response.status !== 404) {
                throw new Error(`Failed to delete shard: ${response.statusText}`);
            }
        } catch (error) {
            console.error(`[NodeManager] Failed to delete shard ${shardId}:`, error);
            throw error;
        }
    }

    /**
     * Distribute shards across nodes with replication
     * Each shard is stored on 2 different nodes for fault tolerance
     * This allows recovery even if multiple nodes fail
     */
    async distributeShards(shards: Shard[]): Promise<ChunkPlacement[]> {
        const REPLICATION_FACTOR = 2; // Store each shard on 2 nodes
        const online = this.getOnlineNodes();

        if (online.length < 2) {
            throw new Error(`Not enough online nodes: need at least 2, have ${online.length}`);
        }

        const placements: ChunkPlacement[] = [];

        // Distribute each shard to multiple nodes
        for (let i = 0; i < shards.length; i++) {
            const shard = shards[i];

            // Select 2 different nodes for this shard
            // Use modulo to spread across nodes, offset by 1 for second replica
            const primaryNodeIndex = i % online.length;
            const secondaryNodeIndex = (i + 1) % online.length;

            const nodesToStore = [
                online[primaryNodeIndex],
                online[secondaryNodeIndex]
            ];

            // Store on both nodes
            for (const node of nodesToStore) {
                try {
                    const placement = await this.storeShard(node, shard);
                    placements.push(placement);
                } catch (error) {
                    console.error(`[NodeManager] Failed to store shard ${i} on ${node.id}, trying next node`);
                    // Continue with other nodes if one fails
                }
            }
        }

        const uniqueNodes = new Set(placements.map(p => p.nodeId)).size;
        console.log(`[NodeManager] Distributed ${shards.length} shards (${placements.length} copies) across ${uniqueNodes} nodes`);
        console.log(`[NodeManager] Replication factor: ${REPLICATION_FACTOR}x - can tolerate ${Math.min(shards.length, uniqueNodes - 1)} node failures`);

        return placements;
    }

    /**
     * Collect shards from nodes with replica fallback
     * With replication, each shard may exist on multiple nodes
     * Try each replica until we get the data
     */
    async collectShards(placements: ChunkPlacement[]): Promise<(Shard | null)[]> {
        // Group placements by shard index (since each shard has replicas)
        const shardPlacements: Map<number, ChunkPlacement[]> = new Map();
        for (const placement of placements) {
            const existing = shardPlacements.get(placement.shardIndex) || [];
            existing.push(placement);
            shardPlacements.set(placement.shardIndex, existing);
        }

        // Determine max shard index
        const maxIndex = Math.max(...Array.from(shardPlacements.keys())) + 1;
        const shards: (Shard | null)[] = new Array(maxIndex).fill(null);
        let retrievedCount = 0;
        let failedNodes: string[] = [];

        // For each unique shard, try to retrieve from any available replica
        await Promise.all(Array.from(shardPlacements.entries()).map(async ([shardIndex, replicas]) => {
            for (const placement of replicas) {
                try {
                    const data = await this.retrieveShard(placement.nodeUrl, placement.chunkId);
                    shards[shardIndex] = {
                        id: placement.chunkId,
                        index: shardIndex,
                        isData: shardIndex < 4, // RS(4,2)
                        data,
                        size: data.length
                    };
                    retrievedCount++;
                    console.log(`[NodeManager] ✓ Retrieved shard ${shardIndex} from ${placement.nodeId}`);
                    return; // Got the shard, no need to try other replicas
                } catch (error) {
                    console.warn(`[NodeManager] ✗ Failed to get shard ${shardIndex} from ${placement.nodeId}, trying replica...`);
                    failedNodes.push(placement.nodeId);
                }
            }
            // If we get here, all replicas failed for this shard
            console.error(`[NodeManager] ✗ All replicas failed for shard ${shardIndex}`);
        }));

        console.log(`[NodeManager] Retrieved ${retrievedCount}/${maxIndex} unique shards`);
        if (failedNodes.length > 0) {
            console.log(`[NodeManager] Failed nodes: ${[...new Set(failedNodes)].join(', ')}`);
        }

        return shards;
    }

    /**
     * Add event listener
     */
    on(callback: (event: string, data: any) => void): void {
        this.listeners.push(callback);
    }

    /**
     * Emit event
     */
    private emit(event: string, data: any): void {
        this.listeners.forEach(cb => cb(event, data));
    }

    /**
     * Get cluster status summary
     */
    getClusterStatus(): {
        totalNodes: number;
        onlineNodes: number;
        offlineNodes: number;
        totalChunks: number;
        totalStorage: number;
        healthy: boolean;
    } {
        const all = this.getAllNodes();
        const online = this.getOnlineNodes();

        return {
            totalNodes: all.length,
            onlineNodes: online.length,
            offlineNodes: all.length - online.length,
            totalChunks: online.reduce((sum, n) => sum + n.chunksStored, 0),
            totalStorage: online.reduce((sum, n) => sum + n.storageUsed, 0),
            healthy: online.length >= 4 // Need at least k nodes for RS(4,2)
        };
    }
}

// Export singleton
export const nodeManager = new NodeManager();

// Export class for testing
export { NodeManager };

File: backend/gateway/services/node_manager.py
"""
Node Manager Service

Manages the cluster of storage nodes (MinIO):
- Tracks node health via heartbeats
- Distributes shards across nodes
- Handles node failures and re-replication
- Provides node selection for uploads
"""

import asyncio
import aiohttp
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Dict, Optional, Callable, Any
import logging

from .erasure_service import Shard

logger = logging.getLogger(__name__)

# Configuration
HEARTBEAT_INTERVAL = 10  # seconds
NODE_TIMEOUT = 30  # seconds before marking offline
REPLICATION_FACTOR = 2  # Store each shard on 2 nodes


@dataclass
class StorageNode:
    """Represents a storage node in the cluster"""
    id: str
    url: str
    port: int
    status: str = "offline"  # online, offline, degraded
    last_heartbeat: datetime = field(default_factory=lambda: datetime.min)
    chunks_stored: int = 0
    storage_used: int = 0
    latency: float = 0.0  # milliseconds


@dataclass
class ChunkPlacement:
    """Tracks where a chunk/shard is stored"""
    chunk_id: str
    shard_index: int
    node_id: str
    node_url: str
    stored: bool = False
    verified: bool = False


class NodeManager:
    """Orchestrates storage cluster operations"""

    def __init__(self):
        self.nodes: Dict[str, StorageNode] = {}
        self._monitoring = False
        self._monitor_task: Optional[asyncio.Task] = None
        self.listeners: List[Callable[[str, Any], None]] = []
        logger.info("[NodeManager] Initialized")

    def register_node(self, node_id: str, url: str, port: int) -> None:
        """Register a storage node"""
        node = StorageNode(
            id=node_id,
            url=url,
            port=port,
            status="offline",
            last_heartbeat=datetime.min
        )
        self.nodes[node_id] = node
        logger.info(f"[NodeManager] Registered node: {node_id} at {url}:{port}")

    def register_minio_nodes(self, minio_endpoints: List[Dict[str, Any]]) -> None:
        """Register MinIO nodes from configuration"""
        for endpoint in minio_endpoints:
            self.register_node(
                node_id=endpoint.get("id", f"minio-{endpoint['port']}"),
                url=endpoint.get("url", "http://localhost"),
                port=endpoint.get("port", 9000)
            )

    def register_default_nodes(self) -> None:
        """Register default MinIO nodes for development (4-node cluster)"""
        # Based on docker-compose configuration
        base_port = 9000
        for i in range(1, 5):  # minio1 to minio4
            self.register_node(
                node_id=f"minio{i}",
                url="http://localhost",
                port=base_port
            )
        logger.info("[NodeManager] Registered 4 default MinIO nodes")

    async def start_monitoring(self) -> None:
        """Start monitoring node health"""
        if self._monitoring:
            return

        self._monitoring = True
        logger.info("[NodeManager] Starting health monitoring")
        await self._check_all_nodes()

        self._monitor_task = asyncio.create_task(self._monitoring_loop())

    async def stop_monitoring(self) -> None:
        """Stop health monitoring"""
        self._monitoring = False
        if self._monitor_task:
            self._monitor_task.cancel()
            try:
                await self._monitor_task
            except asyncio.CancelledError:
                pass
        logger.info("[NodeManager] Monitoring stopped")

    async def _monitoring_loop(self) -> None:
        """Background loop for health checks"""
        while self._monitoring:
            await asyncio.sleep(HEARTBEAT_INTERVAL)
            await self._check_all_nodes()

    async def _check_all_nodes(self) -> None:
        """Check health of all nodes"""
        tasks = [self._check_node(node) for node in self.nodes.values()]
        await asyncio.gather(*tasks, return_exceptions=True)

    async def _check_node(self, node: StorageNode) -> None:
        """Check single node health via MinIO health endpoint"""
        start_time = datetime.now()

        try:
            async with aiohttp.ClientSession() as session:
                # MinIO health check endpoint
                url = f"{node.url}:{node.port}/minio/health/live"
                async with session.get(url, timeout=aiohttp.ClientTimeout(total=5)) as response:
                    if response.status == 200:
                        latency = (datetime.now() - start_time).total_seconds() * 1000
                        was_offline = node.status == "offline"

                        node.status = "online"
                        node.last_heartbeat = datetime.now()
                        node.latency = latency

                        if was_offline:
                            logger.info(f"[NodeManager] Node {node.id} came online ({latency:.0f}ms)")
                            self._emit("node:online", {"node_id": node.id})
                    else:
                        self._mark_node_offline(node)
        except Exception as e:
            logger.debug(f"[NodeManager] Health check failed for {node.id}: {e}")
            self._mark_node_offline(node)

    def _mark_node_offline(self, node: StorageNode) -> None:
        """Mark a node as offline"""
        if node.status != "offline":
            logger.warning(f"[NodeManager] Node {node.id} is OFFLINE")
            node.status = "offline"
            self._emit("node:offline", {"node_id": node.id})

    def get_online_nodes(self) -> List[StorageNode]:
        """Get list of online nodes"""
        return [n for n in self.nodes.values() if n.status == "online"]

    def get_all_nodes(self) -> List[StorageNode]:
        """Get all registered nodes"""
        return list(self.nodes.values())

    def select_nodes_for_shards(self, shard_count: int) -> List[StorageNode]:
        """
        Select nodes for storing shards.
        Uses least-loaded preference with latency as tiebreaker.
        """
        online = self.get_online_nodes()

        if len(online) < shard_count:
            raise ValueError(f"Not enough online nodes: need {shard_count}, have {len(online)}")

        # Sort by storage used (least loaded first) then by latency
        sorted_nodes = sorted(online, key=lambda n: (n.storage_used, n.latency))
        return sorted_nodes[:shard_count]

    async def store_shard(self, node: StorageNode, shard: Shard, minio_client) -> ChunkPlacement:
        """Store a shard on a specific node using MinIO client"""
        from io import BytesIO

        bucket_name = "file-shards"
        object_name = f"shard_{shard.id}"

        try:
            # Ensure bucket exists
            if not minio_client.bucket_exists(bucket_name):
                minio_client.make_bucket(bucket_name)

            # Upload shard data
            data_stream = BytesIO(shard.data)
            minio_client.put_object(
                bucket_name,
                object_name,
                data_stream,
                length=shard.size,
                metadata={
                    "shard_index": str(shard.index),
                    "is_data": str(shard.is_data),
                    "node_id": node.id
                }
            )

            logger.info(f"[NodeManager] Stored shard {shard.id[:8]}... on {node.id}")

            return ChunkPlacement(
                chunk_id=shard.id,
                shard_index=shard.index,
                node_id=node.id,
                node_url=f"{node.url}:{node.port}",
                stored=True,
                verified=True
            )
        except Exception as e:
            logger.error(f"[NodeManager] Failed to store shard on {node.id}: {e}")
            raise

    async def retrieve_shard(self, node_url: str, shard_id: str, minio_client) -> bytes:
        """Retrieve a shard from MinIO"""
        bucket_name = "file-shards"
        object_name = f"shard_{shard_id}"

        try:
            response = minio_client.get_object(bucket_name, object_name)
            data = response.read()
            response.close()
            response.release_conn()
            return data
        except Exception as e:
            logger.error(f"[NodeManager] Failed to retrieve shard {shard_id}: {e}")
            raise

    async def delete_shard(self, shard_id: str, minio_client) -> None:
        """Delete a shard from MinIO"""
        bucket_name = "file-shards"
        object_name = f"shard_{shard_id}"

        try:
            minio_client.remove_object(bucket_name, object_name)
        except Exception as e:
            logger.error(f"[NodeManager] Failed to delete shard {shard_id}: {e}")
            raise

    async def distribute_shards(
        self,
        shards: List[Shard],
        minio_client
    ) -> List[ChunkPlacement]:
        """
        Distribute shards across nodes with replication.
        Each shard is stored on multiple nodes for fault tolerance.
        """
        online = self.get_online_nodes()

        if len(online) < 2:
            raise ValueError(f"Not enough online nodes: need at least 2, have {len(online)}")

        placements = []

        for i, shard in enumerate(shards):
            # Select nodes for this shard (primary and secondary for replication)
            primary_idx = i % len(online)
            secondary_idx = (i + 1) % len(online)

            nodes_to_store = [online[primary_idx]]
            if REPLICATION_FACTOR > 1 and len(online) > 1:
                nodes_to_store.append(online[secondary_idx])

            # Store on selected nodes
            for node in nodes_to_store:
                try:
                    placement = await self.store_shard(node, shard, minio_client)
                    placements.append(placement)
                except Exception as e:
                    logger.warning(f"[NodeManager] Failed to store shard {i} on {node.id}: {e}")

        unique_nodes = len(set(p.node_id for p in placements))
        logger.info(f"[NodeManager] Distributed {len(shards)} shards ({len(placements)} copies) across {unique_nodes} nodes")

        return placements

    async def collect_shards(
        self,
        placements: List[ChunkPlacement],
        minio_client
    ) -> List[Optional[Shard]]:
        """
        Collect shards from nodes with replica fallback.
        Each shard may exist on multiple nodes - try each until successful.
        """
        # Group placements by shard index
        shard_placements: Dict[int, List[ChunkPlacement]] = {}
        for p in placements:
            if p.shard_index not in shard_placements:
                shard_placements[p.shard_index] = []
            shard_placements[p.shard_index].append(p)

        # Determine max shard index
        if not shard_placements:
            return []

        max_index = max(shard_placements.keys()) + 1
        shards: List[Optional[Shard]] = [None] * max_index

        # Collect each shard
        for shard_index, replicas in shard_placements.items():
            for placement in replicas:
                try:
                    data = await self.retrieve_shard(
                        placement.node_url,
                        placement.chunk_id,
                        minio_client
                    )
                    shards[shard_index] = Shard(
                        id=placement.chunk_id,
                        index=shard_index,
                        is_data=shard_index < 4,  # RS(4,2)
                        data=data,
                        size=len(data)
                    )
                    logger.info(f"[NodeManager] Retrieved shard {shard_index} from {placement.node_id}")
                    break  # Got the shard, no need for other replicas
                except Exception as e:
                    logger.warning(f"[NodeManager] Failed to get shard {shard_index} from {placement.node_id}: {e}")

        retrieved = len([s for s in shards if s is not None])
        logger.info(f"[NodeManager] Retrieved {retrieved}/{max_index} shards")

        return shards

    def on(self, callback: Callable[[str, Any], None]) -> None:
        """Add event listener"""
        self.listeners.append(callback)

    def _emit(self, event: str, data: Any) -> None:
        """Emit event to all listeners"""
        for callback in self.listeners:
            try:
                callback(event, data)
            except Exception as e:
                logger.error(f"[NodeManager] Event listener error: {e}")

    def get_cluster_status(self) -> Dict[str, Any]:
        """Get cluster status summary"""
        all_nodes = self.get_all_nodes()
        online = self.get_online_nodes()

        return {
            "total_nodes": len(all_nodes),
            "online_nodes": len(online),
            "offline_nodes": len(all_nodes) - len(online),
            "total_chunks": sum(n.chunks_stored for n in online),
            "total_storage": sum(n.storage_used for n in online),
            "healthy": len(online) >= 4  # Need at least k nodes for RS(4,2)
        }


# Singleton instance
node_manager = NodeManager()
